src/codegen/mod.rs

```diff
@@ -127,55 +132,133 @@ fn gen_stmt(
+                let (init_expr, cu_ty, checks) = match gened_ty {
                     cu::Ty::Array(_, _) => {
-                        (gen_expr(e1, parall_ctx, view_ctx, comp_unit), gened_ty)
+                        let (ex, ch) = match gen_expr(e1, parall_ctx, view_ctx, comp_unit, label.clone()) {
+                            CheckedExpr::Expr(e) => (e, None),
+                            CheckedExpr::ExprIdxCheck(c, e) => (e, Some(c)),
+                        };
+                        (ex, gened_ty, ch)
+                    }
+                    _ => {
+                        let (ex, ch) = match gen_expr(e1, parall_ctx, view_ctx, comp_unit, label.clone()) {
+                            CheckedExpr::Expr(e) => (e, None),
+                            CheckedExpr::ExprIdxCheck(c, e) => (e, Some(c)),
+                        };
+                        (
+                            ex,
+                            if *mutbl == desc::Mutability::Mut {
+                                cu::Ty::Scalar(cu::ScalarTy::Auto)
+                            } else {
+                                cu::Ty::Const(Box::new(cu::Ty::Scalar(cu::ScalarTy::Auto)))
+                            },
+                            ch,
+                        )
                     }
-                    _ => (
-                        gen_expr(e1, parall_ctx, view_ctx, comp_unit),
-                        if *mutbl == desc::Mutability::Mut {
-                            cu::Ty::Scalar(cu::ScalarTy::Auto)
-                        } else {
-                            cu::Ty::Const(Box::new(cu::Ty::Scalar(cu::ScalarTy::Auto)))
-                        },
-                    ),
                 };
                 let var = cu::Stmt::VarDecl {
                     name: ident.name.clone(),
                     ty: cu_ty,
                     addr_space: None,
                     expr: Some(init_expr),
                 };
-                cu::Stmt::Seq(
-                    Box::new(var),
-                    Box::new(gen_stmt(e2, return_value, parall_ctx, view_ctx, comp_unit)),
-                )
+                match (
+                    checks,
+                    gen_stmt(e2, return_value, parall_ctx, view_ctx, comp_unit, label),
+                ) {
+                    (None, CheckedStmt::Stmt(st)) => {
+                        CheckedStmt::Stmt(cu::Stmt::Seq(Box::new(var), Box::new(st)))
+                    }
+                    (Some(ch), CheckedStmt::Stmt(st)) => CheckedStmt::Stmt(cu::Stmt::Seq(
+                        Box::new(ch),
+                        Box::new(cu::Stmt::Seq(Box::new(var), Box::new(st))),
+                    )),
+                    (None, CheckedStmt::StmtIdxCheck(ch, st)) => CheckedStmt::Stmt(cu::Stmt::Seq(
+                        Box::new(var),
+                        Box::new(cu::Stmt::Seq(Box::new(ch), Box::new(st))),
+                    )),
+                    (Some(ch1), CheckedStmt::StmtIdxCheck(ch2, st)) => {
```

Comment:
having sequences inside of sequences happens a lot while generating checks. I would implement a function like `fn build_seq(vec<cu::Stmt>) -> cu::Stmt`.